### PR TITLE
New receipe / first step

### PR DESCRIPTION
**Pull request by AnastasiaHolovash**

### CODE REVIEW

File: DishBookIOS/Classes/Coordinators/Common/AppCoordinator.swift [hunk 1]
```diff
@@ -11,6 +11,7 @@ final class AppCoordinator: BaseCoordinator {
     
     let window: UIWindow
     
+
```
Comment (ddanilyuk) [inline DishBookIOS/Classes/Coordinators/Common/AppCoordinator.swift]: Unneed spacer

File: DishBookIOS/Classes/Coordinators/MainFlow/NewDishCoordinator.swift [hunk 2]
```diff
@@ -6,23 +6,65 @@
...
 
 import UIKit
+import Combine
 
 final class NewDishCoordinator: BaseRootCoordinator {
-        
+    
+    private var newDish = NewDish()
+    private var cancelableSet: Set<AnyCancellable> = []
+    let viewModel = NewDishViewModel()
```
Comment (ddanilyuk) [inline DishBookIOS/Classes/Coordinators/MainFlow/NewDishCoordinator.swift]: Remove from coordinator

File: DishBookIOS/Classes/Coordinators/MainFlow/NewDishCoordinator.swift [hunk 3]
```diff
@@ -6,23 +6,65 @@
...
+    lazy var firstStepViewController = NewDishViewController(viewModel: viewModel)
```
Comment (ddanilyuk) [inline DishBookIOS/Classes/Coordinators/MainFlow/NewDishCoordinator.swift]: Rename NewDishViewController -> FirstStepViewComtroller

**Review by ddanilyuk — CHANGES_REQUESTED** [review]

File: DishBookIOS/Classes/Coordinators/MainFlow/NewDishCoordinator.swift [hunk 4]
```diff
@@ -6,23 +6,63 @@
 //
 
 import UIKit
+import Combine
 
 final class NewDishCoordinator: BaseRootCoordinator {
-        
+    
+    private var newDish = NewDish()
+    private var cancelableSet: Set<AnyCancellable> = []
```
Comment (ddanilyuk) [inline DishBookIOS/Classes/Coordinators/MainFlow/NewDishCoordinator.swift]: Maybe add this variable to BaseFlowCoordinator or something like this?

File: DishBookIOS/Classes/Models/Dish.swift [hunk 5]
```diff
@@ -80,6 +80,10 @@ extension Dish {
 
 // MARK: - Nested types
 
+protocol StringConvertible {
```
Comment (ddanilyuk) [inline DishBookIOS/Classes/Models/Dish.swift]: Move to extensions

File: DishBookIOS/Classes/Modules/NewDish/ViewModel/FirstStepViewModel.swift [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+//
+//  Created by Denys Danyliuk on 02.02.2021.
+//
+
+import Foundation
+import Combine
+
+final class FirstStepViewModel: BaseViewModel {
+    
+    // Publishers
```
Comment (ddanilyuk) [inline DishBookIOS/Classes/Modules/NewDish/ViewModel/FirstStepViewModel.swift]: Add `Mark: -`

File: DishBookIOS/Classes/Modules/NewDish/ViewModel/FirstStepViewModel.swift [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    lazy var didSelectPrivacyLevelPublisher = didSelectPrivacyLevelSubject.eraseToAnyPublisher()
+    lazy var didSelectDifficultyLevelPublisher = didSelectDifficultyLevelSubject.eraseToAnyPublisher()
+    lazy var didChangeNamePublisher = didChangeNameSubject.eraseToAnyPublisher()
+    lazy var didChangeNumberOfServingsPublisher = didChangeNumberOfServingsSubject.eraseToAnyPublisher()
+    lazy var didPressNextPublisher = didPressNextSubject.eraseToAnyPublisher()
+    
+    // Subjects
```
Comment (ddanilyuk) [inline DishBookIOS/Classes/Modules/NewDish/ViewModel/FirstStepViewModel.swift]: Here too

File: DishBookIOS/Classes/Modules/NewDish/ViewModel/NewDish.swift [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+struct NewDish {
+    
+    /// Document ID in `Firestore Database`
+    @DocumentID var id: String?
+    
+    /// Property used when dish was added by another user
+    /// It contains the id of the document in `Public dishes` collection
+    /// If this property is not nil, `userAddedID` property
+    /// must be not nil too.
+    var dishID: String?
```
Comment (ddanilyuk) [inline DishBookIOS/Classes/Modules/NewDish/ViewModel/NewDish.swift]: New dish do not have this property

File: DishBookIOS/Classes/Modules/NewDish/ViewModel/NewDish.swift [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+// MARK: - Hashable
+
+extension NewDish: Hashable {
+    
+}
+
+// MARK: - Codable
+
+//extension NewFakingDish: Codable {
```
Comment (ddanilyuk) [inline DishBookIOS/Classes/Modules/NewDish/ViewModel/NewDish.swift:56]: Remove comments

File: DishBookIOS/Classes/Views/AddDish/SegmentedControl.swift [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+    // MARK: - @objc
+    
+    @objc
+    func sliderTapped(_ gestureRecognizer: UIGestureRecognizer) {
+        
+        let pointTapped: CGPoint = gestureRecognizer.location(in: self)
+        let index = Int(pointTapped.x / segmentWidth)
+        
+        self.selectedView.apply(style: Styles.View.CornerRadius.d0)
+        self.selectedView.apply(style: Styles.View.CornerRadius.noMasked)
```
Comment (ddanilyuk) [inline DishBookIOS/Classes/Views/AddDish/SegmentedControl.swift]: Remove self

File: DishBookIOS/Classes/Views/AddDish/SegmentedControl.swift [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+        selectedLabel.font = R.font.sfProRoundedSemibold(size: 16)
+        
+        UIView.animate(withDuration: 0.6, delay: 0, usingSpringWithDamping: 1, initialSpringVelocity: 0.5, options: UIView.AnimationOptions.curveEaseIn) {
```
Comment (ddanilyuk) [inline DishBookIOS/Classes/Views/AddDish/SegmentedControl.swift]: Add `[self] in` and remove self inside closure

File: DishBookIOS/Classes/Views/AddDish/SegmentedControl.swift [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+        super.beginTracking(touch, with: event)
+        
+        return true
+    }
+    override func continueTracking(_ touch: UITouch, with event: UIEvent?) -> Bool {
+        super.continueTracking(touch, with: event)
+        
+        return true
+    }
+    override func endTracking(_ touch: UITouch?, with event: UIEvent?) {
```
Comment (ddanilyuk) [inline DishBookIOS/Classes/Views/AddDish/SegmentedControl.swift]: Unused method

File: DishBookIOS/Classes/Views/AddDish/SegmentedControl.swift [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+            selectedLabel.font = R.font.sfProRoundedSemibold(size: 15)
+            didSelectSubject.send(segments[index])
+        }
+    }
+    
+    
+    override func beginTracking(_ touch: UITouch, with event: UIEvent?) -> Bool {
+        super.beginTracking(touch, with: event)
+        
+        return true
```
Comment (ddanilyuk) [inline DishBookIOS/Classes/Views/AddDish/SegmentedControl.swift]: Maybe it's unused too

File: DishBookIOS/Classes/Views/AddDish/SteperView.swift [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+        button.apply(style: Styles.View.CornerRadius.d17)
+        button.titleEdgeInsets = UIEdgeInsets(top: 0, left: 1, bottom: 4, right: 0)
+        button.backgroundColor = .white
+        button.setTitle(title, for: .normal)
+        button.setTitleColor(R.color.textBlack(), for: .normal)
+        button.titleLabel?.font = R.font.sfProRoundedMedium(size: 30)
+    }
+}
+
+import SwiftUI
```
Comment (ddanilyuk) [inline DishBookIOS/Classes/Views/AddDish/SteperView.swift:135]: Maybe move this block with preview to #if DEBUG

**Review by ddanilyuk — APPROVED** [review]

File: DishBookIOS/Classes/Coordinators/MainFlow/NewDishCoordinator.swift (hunk 4 above)
Comment (AnastasiaHolovash) [reply DishBookIOS/Classes/Coordinators/MainFlow/NewDishCoordinator.swift]: cancelableSet?

Comment (ddanilyuk) [reply DishBookIOS/Classes/Coordinators/MainFlow/NewDishCoordinator.swift]: Yes

Comment (ddanilyuk) [reply DishBookIOS/Classes/Coordinators/MainFlow/NewDishCoordinator.swift]: Like in BaseViewModel and BaseViewController

File: DishBookIOS/Classes/Modules/NewDish/ViewModel/FirstStepViewModel.swift (hunk 6 above)
Comment (ddanilyuk) [reply DishBookIOS/Classes/Modules/NewDish/ViewModel/FirstStepViewModel.swift]: And after mark must be empty line

File: DishBookIOS/Classes/Views/AddDish/SteperView.swift (hunk 14 above)
Comment (AnastasiaHolovash) [reply DishBookIOS/Classes/Views/AddDish/SteperView.swift:135]: I think it's not necessary 
we don't need two additional lines of code in every view)

**Pull request merged**